chirp/drivers/tdh8.py

```diff
@@ -0,0 +1,2222 @@
+
+    def get_features(self):
+        rf = chirp_common.RadioFeatures()
+        rf.has_settings = True
+        rf.has_bank = False
+        rf.has_cross = True
+        rf.has_ctone = True
+        rf.has_rx_dtcs = True
+        rf.has_tuning_step = False
+        rf.has_ctone = True
+        rf.can_odd_split = True
+        rf.valid_name_length = 8
+        rf.valid_characters = self._valid_chars
+        rf.valid_skips = []
+        rf.valid_tmodes = ["", "Tone", "TSQL", "DTCS", "Cross"]
+        rf.valid_cross_modes = [
+            "Tone->Tone",
+            "DTCS->",
+            "->DTCS",
+            "Tone->DTCS",
+            "DTCS->Tone",
+            "->Tone",
+            "DTCS->DTCS"]
+        rf.valid_power_levels = TX_POWER
+        rf.valid_duplexes = ["", "-", "+", "split", "Off"]
+        rf.valid_modes = ["FM", "NFM"]
+        rf.valid_tuning_steps = STEPS
+
+        normal_bands = [self._vhf_range, self._uhf_range]
+
+        if self._mmap is None:
+            rf.valid_bands = [normal_bands[0], normal_bands[1]]
+        else:
+            rf.valid_bands = normal_bands
+        rf.memory_bounds = (1, 199)
+        return rf
+
+    def process_mmap(self):
+        self._memobj = bitwise.parse(MEM_FORMAT, self._mmap)
+
+    def validate_memory(self, mem):
+        msgs = chirp_common.CloneModeRadio.validate_memory(self, mem)
+
+        _msg_duplex2 = 'Memory location only supports "(None)" or "off"'
+        _msg_duplex3 = 'Memory location only supports "(None)", "+" or "off"'
+
+        if self._gmrs:
+            if mem.number < 1 or mem.number > 30:
+                if float(mem.freq) / 1000000 in GMRS_FREQS1:
+                    if mem.duplex not in ['', 'off']:
+                        # warn user wrong Duplex
+                        msgs.append(chirp_common.ValidationError(_msg_duplex2))
+
+                if float(mem.freq) / 1000000 in GMRS_FREQS2:
+                    if mem.duplex not in ['', 'off']:
+                        # warn user wrong Duplex
+                        msgs.append(chirp_common.ValidationError(_msg_duplex2))
+
+                if float(mem.freq) / 1000000 in GMRS_FREQS3:
+                    if mem.duplex not in ['', '+', 'off']:
+                        # warn user wrong Duplex
+                        msgs.append(chirp_common.ValidationError(_msg_duplex3))
+
+        return msgs
+
+    def sync_in(self):
+        try:
+            self._mmap = _do_download(self)
+        except Exception as e:
+            raise errors.RadioError("Failed to communicate with radio: %s" % e)
+
+        self.process_mmap()
+        if self.radio_ident == "P31183":
+            self._memobj.settings.ham = 0
+            self._memobj.settings.gmrs = 0
+        elif self.radio_ident == "P31184":
+            self._memobj.settings.ham = 0
+            self._memobj.settings.gmrs = 1
+        elif self.radio_ident == "P31185":
+            self._memobj.settings.ham = 1
+            self._memobj.settings.gmrs = 0
```

Comment:
In all three of these cases, why are you modifying the image as soon as you download it? These values should already be set in the radio's memory if you've confirmed that the user selected the right model, correct? It is very unconventional for chirp to modify the image after download, as saving the result of a radio download should exactly match what is in the radio.